### PR TITLE
Added AttrList

### LINKED ISSUE #83: Add text style attributes support
We need at least:
- [ ] `pango_attr_list_new`
- [ ] `pango_attr_list_insert`
- [ ] `pango_layout_set_attributes`

### PR DESCRIPTION
**Pull request by EPashkin**
Closes #83

### CODE REVIEW

**Comment by EPashkin** [comment]
@RazrFalcon can you test if this works?
And can you give snipped that can used as standalone example (I don't know how use pango right way)?

**Comment by GuillaumeGomez** [comment]
Damn, you did it before me!

**Comment by RazrFalcon** [comment]
>I don't know how use pango right way?
Neither do I. I'm still learning.

How can I build this branch?

**Comment by EPashkin** [comment]
https://help.github.com/articles/checking-out-pull-requests-locally/

**Comment by RazrFalcon** [comment]
Strange. I've tried this code:
```rust
let attr_list = pango::AttrList::new();
let attr = pango::functions::attr_underline_new(pango::Underline::Single).unwrap();
attr_list.insert(&attr);
layout.set_attributes(Some(&attr_list));
```

and I got:
```
thread 'main' panicked at 'not yet implemented', /home/razr/.cargo/git/checkouts/glib-928cf7b282977403/f874842/src/translate.rs:262
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:371
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:511
   6: glib::translate::ToGlibPtr::to_glib_full
             at ./<panic macros>:3
   7: pango::attr_list::<impl pango::auto::attr_list::AttrList>::insert
             at /media/data/Programming/Sources/pango/src/attr_list.rs:19
...
```

**Comment by EPashkin** [comment]
I was afraid of this: no to_glib_full for Boxed.
Try to fix it

**Comment by EPashkin** [comment]
Updated

**Comment by RazrFalcon** [comment]
It works!

**Comment by RazrFalcon** [comment]
Also a warning:
```
 --> pango/src/auto/engine_shape.rs:5:5
  |
5 | use glib::translate::*;
  |     ^^^^^^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default
``

**Comment by GuillaumeGomez** [comment]
Don't worry about warnings in generated code. ;)

**Comment by EPashkin** [comment]
@RazrFalcon can you link your current code, I try extract example from it.

**Comment by EPashkin** [comment]
@GuillaumeGomez, @sdroege Please, check manual functions in `attr_list.rs`.
I don't like current version, but can't find any other variants, maybe you have idea.

**Comment by RazrFalcon** [comment]
@EPashkin I will try to create a minimal example tomorrow. Current code is from my project.

**Comment by GuillaumeGomez** [comment]
@EPashkin: Unfortunately, there aren't many ways to avoid rust running the destructor once the variable goes out of scope. For me, it's good enough and the code is really easy to understand so there's no point (from my point of view) in trying to find something else for the moment. Good job.

**Comment by RazrFalcon** [comment]
Cargo.toml
```toml
[package]
name = "pango-test"
version = "0.1.0"

[dependencies.pango]
# git = "https://github.com/gtk-rs/pango"
# custom branch
path = "../pango" 

[dependencies.pangocairo]
# git = "https://github.com/RazrFalcon/pangocairo-rs"
# with updated dependencies 
path = "../pangocairo"

[dependencies.cairo-rs]
git = "https://github.com/gtk-rs/cairo"
features = ["png"]
```

main.rs
```rust
extern crate pango;
extern crate cairo;
extern crate pangocairo;

use std::fs;

use pango::LayoutExt;
use pangocairo::{PangoContextExt, CairoContextExt};

fn main() {
    let s = cairo::ImageSurface::create(cairo::Format::ARgb32, 200, 200);
    let cr = cairo::Context::new(&s);

    cr.set_source_rgba(1.0, 1.0, 1.0, 1.0);
    cr.paint();

    let pc = cr.create_pango_context();
    pc.set_resolution(300.0);
    cr.update_pango_context(&pc);

    let mut font = pango::FontDescription::new();
    font.set_family("Verdana");
    font.set_size(14 * pango::SCALE);

    let layout = pango::Layout::new(&pc);
    layout.set_font_description(Some(&font));
    layout.set_text("Text", -1);

    let attr_list = pango::AttrList::new();
    let attr = pango::functions::attr_underline_new(pango::Underline::Single).unwrap();
    attr_list.insert(attr);
    layout.set_attributes(Some(&attr_list));

    cr.move_to(10.0, 0.0);
    cr.set_source_rgba(1.0, 0.0, 0.0, 1.0);

    cr.update_pango_layout(&layout);
    cr.show_pango_layout(&layout);

    cr.pango_layout_path(&layout);
    cr.stroke();

    let mut buffer = fs::File::create("out.png").unwrap();
    s.write_to_png(&mut buffer).unwrap();
}
```

Hope it would help.

Result:
out

**Comment by EPashkin** [comment]
@RazrFalcon Thanks, I tested getter/setters on it.
Bad that it don't works without pangocairo.

**Comment by RazrFalcon** [comment]
@EPashkin You need pangocairo to draw text on something. I don't know is it possible to draw text using only pango. It's just a layout engine after all.

**Comment by RazrFalcon** [comment]
@EPashkin Also `pango::functions::attr_underline_new()` looks super ugly. Can we do something about it? Should I create an issue for this?

**Comment by EPashkin** [comment]
You can use `pango::attr_underline_new`.
Or you want something other?

**Comment by RazrFalcon** [comment]
I want something like: `Attribute::underline(Underline::Single)`. Implement this methods as static methods of `Attribute`.

**Comment by EPashkin** [comment]
Yes, I can do it in this PR.
Only problem with names: IMHO better `new_underline` instead `attr_underline_new` to be clear.

**Comment by RazrFalcon** [comment]
Good. `new_underline` looks more like idiomatic rust.

**Comment by EPashkin** [comment]
Updated

**Comment by EPashkin** [comment]
Seems this ready to merge

**Comment by GuillaumeGomez** [comment]
I agree. Thanks!

File: src/attr_list.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+use Attribute;
+use AttrList;
+use ffi;
+use glib::translate::*;
+use std::mem;
+
+impl AttrList {
+    pub fn change(&self, attr: Attribute) {
+        unsafe {
+            ffi::pango_attr_list_change(self.to_glib_none().0, attr.to_glib_none().0 as *mut _);
```
Comment (sdroege) [inline src/attr_list.rs]: That's exactly why I'd like to have https://github.com/gtk-rs/glib/issues/176 :)

**Comment by sdroege** [comment]
Looks good to me too

**Pull request merged**